database/src/service/read.rs

```diff
@@ -47,145 +49,163 @@ type ResponseSend = tokio::sync::oneshot::Sender<ResponseResult>;
-    pub(super) sender: crossbeam::channel::Sender<(ReadRequest, ResponseSend)>,
+    /// This spawns `N` amount of `DatabaseReader`'s
+    /// attached to `db` and returns a handle to the pool.
+    ///
+    /// Should be called _once_ per actual database.
+    #[cold]
+    #[inline(never)] // Only called once.
+    pub(super) fn init(db: &Arc<ConcreteEnv>, reader_threads: ReaderThreads) -> Self {
+        // How many reader threads to spawn?
+        let reader_count = reader_threads.as_threads();
+
+        // Spawn `rayon` reader threadpool.
+        let pool = rayon::ThreadPoolBuilder::new()
+            .num_threads(reader_count.get())
+            .thread_name(|i| format!("cuprate_helper::service::read::DatabaseReader{i}"))
+            .build()
+            .unwrap();
+
+        // Return a handle to the pool.
+        Self {
+            pool: Arc::new(pool),
+            db: Arc::clone(db),
+        }
+    }
+
+    /// TODO
+    #[inline]
+    pub const fn db(&self) -> &Arc<ConcreteEnv> {
+        &self.db
+    }
 }
 
 impl tower::Service<ReadRequest> for DatabaseReadHandle {
     type Response = Response;
     type Error = RuntimeError;
-    type Future = ResponseRecv;
+    type Future = ResponseReceiver;
 
     #[inline]
     fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        todo!()
+        // Threadpool is always ready as long as this handle is alive.
+        Poll::Ready(Ok(()))
     }
 
-    #[inline]
-    fn call(&mut self, _req: ReadRequest) -> Self::Future {
-        todo!()
+    fn call(&mut self, request: ReadRequest) -> Self::Future {
+        // Response channel we `.await` on.
+        let (response_sender, receiver) = oneshot::channel();
+
+        // Database thread's state.
+        let db_reader = DatabaseReader {
+            db: Arc::clone(&self.db),
+        };
+
+        // Spawn the request in the rayon DB thread-pool.
+        //
+        // Note that this uses `install()` instead of `spawn()`
+        // such that any `rayon` parallel code that runs within
+        // the passed closure uses the same `rayon` threadpool.
+        //
+        // INVARIANT:
+        // The below `DatabaseReader` function impl block relies on this behavior.
+        self.pool
+            .install(move || db_reader.map_request(request, response_sender));
+
+        InfallibleOneshotReceiver::from(receiver)
     }
 }
 
 //---------------------------------------------------------------------------------------------------- DatabaseReader
-/// Database reader thread.
+/// Database reader thread state.
 ///
-/// This struct essentially represents a thread.
+/// This struct represents (some of the) state every
+/// `DatabaseReadHandle` rayon thread will have access to.
 ///
-/// Each reader thread is spawned with access to this struct (self).
+/// Could just be function inputs, although this allows `self`
+/// and a `Drop` impl such that each request has the same cleanup routine.
```

Comment:
IMO this would probably be clearer as just function inputs, instead of the extra struct